### DISCUSSION TITLE
How to store refresh_token and access_token

### DISCUSSION THREAD

**Discussion by ConnorVO** [sergiodxa/remix-auth#137] [Q&A]
I know this is outside the scope of this project, but would love some help as I've been banging my head against the wall all day.

I believe you would typically store these using a 'Set-Cookie' header. I've been trying to do that in my `setUserData()` function that I call in the verify callback after I get the tokens and profile. But this just doesn't seem to be keeping those tokens around.

Flow:
- User authenticates and my authenticator strategy calls the verify method
- I send data to a remix resource route 
- In the action for the remix-resource route, I save the data and then try to set the access token and refresh token using 'Set-Cookie' but it doesn't work. 

Including the relevant part of my /api/setUser resource route where I try to set the cookie

```
export let action: ActionFunction = async ({ request }) => {
    const session = await sessionStorage.getSession(
        request.headers.get('Cookie')
    );
    const jsonData = await request.json();
    session.set('access_token', jsonData['accessToken']);
    session.set('refresh_token', jsonData['refreshToken']);

    return json<LoaderData>(
        { status: 'ok' },
        {
            headers: {
                'Set-Cookie': await sessionStorage.commitSession(session),
            },
        }
    );
};
```

**Comment by sergiodxa** [comment]
I can think of three options here:

1. Check if you really need to store the tokens, if you are not going to fetch the API of your provider for something then storing the tokens is useless once you have the profile and all strategies gives you the user profile.
2. You can return them as part of the data the strategy returns and then when you do `authenticator.isAuthenticated(request)` the result object will have the access and refresh token, then when you use them if it fails you can refresh the access token and update the values on the session.
3. You could store them in the DB, with the expiration date, and run a background job to refresh the access token periodically before they expire.

---

In the case you'll fetch the provider API and you do need to tokens I think the third options is the best because the loaders or actions code doesn't need to care about refreshing the token.

You can create a Sessions table on your DB, there store the OAuth2 provider (e.g. GitHub or Google), the user ID, the access token, the refresh token and the expiration date (the time it takes to expire comes on the extraParams, you can use that to get the exact date).

Then, run a background job that uses the refresh token and access token, you can run it every hour (I think most tokens expires after an hour, but check this with each provider), get all the sessions expired or about to expire, check the provider name so you know the endpoints and refresh them, then update the expiration date and tokens on the sessions table.

Now, on your loaders, when you need the access token, you can query that table and get the token and be sure it will be up to date.

**Reply by dan-cooke** [reply]
Hey @sergiodxa thanks for these insights. 

I forsee some issues with the suggested approaches however.

1. Obviously if we don't need to store the token, then this point is the best solution - but in most cases we will be retrieving an access token from our provider in order to query a protected API in our backend.
2. Cookies have a maximum length of 4096 bytes. With accessToken, refreshToken and expires - this will exceed the maximum cookie size from my investigations
3. Storing in a db and updating hourly greatly increases system vulnerability and adds uneccesary resource overhead IMO.

I would be very much interested in some documentation around best practices for storing these values etc. This strategy based approach is new to me, and I'm struggling to arrive on a good solution for storing tokens / profile data in the session
